subsys/bluetooth/host/gatt.c

```diff
@@ -4650,7 +4650,10 @@ int bt_gatt_unsubscribe(struct bt_conn *conn,
 
 void bt_gatt_cancel(struct bt_conn *conn, void *params)
 {
-	bt_att_req_cancel(conn, params);
+	struct bt_att_req *req;
```

Comment:
I think we should explain how canceling works when having multiple requests with the same parameters.
Which order, how many submits.

Reply:
We can specify that the cancel will stop one of the requests (or none is all are sent). Which request is canceled is moot since they are indistinguishable to both the application and the remote peer.

We can be conservative and forbid multiple requests with the same params object. I think this is the simpler contract to explain.

Reply:
Alternatively we can cancel all requests from the parameters if there are multiple.

Reply:
> The list is ordered

Good point. The order in relation to other requests matters. `att->reqs` is in sending order. So the request issued first will be canceled. Do we have a real preference for one behavior over another?

What is the use case for this method? I'm wondering if anybody uses this, since it was broken and nobody noticed. Maybe we could deprecate this?

Reply:
I don't have a preference no.
Deprecating would be an option, it is a while since this was broken, like 2.2/2.3 maybe.

Reply:
I found a new problem with `bt_gatt_cancel`. It does not return any status. So the application does not know if it can free/reuse `params`. I can add a call to `req->func` with an error code. But there is no appropriate error code for "canceled", so it's awkward.

I think `att_handle_rsp` has the same problem. If the ongoing is canceled (`chan->req == &cancel`), `req->func` is never called, so the application is never told to recycle `user_data`/`params`.

Unless I'm confused about the above, I believe it's best to deprecate the canceling feature. At least until we have a concrete use case. @joerchan Could you help me look at whether `bt_gatt_unsubscribe` really needs to call `bt_gatt_cancel`? I have a feeling the design would be cleaner if we just treat unsubscribe as a request and sequence it like any other.

Reply:
You mean you wish to just pend the write to unsubscribe, while the write to subscribe is in progress?
I don't see a problem with that solution. `bt_gatt_subscribe` remove it from the list already, so you'll have to see if that is OK, and how `gatt_write_ccc_rsp` handles that case.

And what would happen if you try to subscribe while unsubscribe is in progress maybe?